src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/jpa/repository/ReferralRepository.kt

```diff
@@ -20,4 +24,73 @@ interface ReferralRepository : JpaRepository<Referral, UUID> {
   // queries for draft referrals
   fun findByIdAndSentAtIsNull(id: UUID): Referral?
   fun findByCreatedByIdAndSentAtIsNull(userId: String): List<Referral>
+
+  @Query(value =
+    "WITH attended_sessions AS ( " +
+        "select count(app.id) AS attended, min(app.appointment_time) as first_appointment, aps.action_plan_id, act.referral_id " +
+            "from appointment app " +
+                "join action_plan_session_appointment apsa on app.id = apsa.appointment_id " +
+                "join action_plan_session aps on apsa.action_plan_session_id = aps.id " +
+                "join action_plan act on aps.action_plan_id = act.id " +
+            "where attended in ('YES', 'LATE') " +
+            "group by aps.action_plan_id, act.referral_id), " +
+        "outcomes_to_be_achieved AS ( " +
+            "select count(des.id) as outcomes_to_be_achieved, ref.id AS referral_id " +
+                "from referral ref " +
+                    "inner join referral_selected_service_category rssc on rssc.referral_id = ref.id " +
+                    "inner join desired_outcome des on des.service_category_id = rssc.service_category_id " +
+                "group by ref.id), " +
+        "achieved_outcomes AS ( " +
+            "select sub.partially_achieved + sub.achieved_count AS total, referral_id " +
+                "from (select ref.id AS referral_id, " +
+                    "sum(case when eosro.achievement_level = 'PARTIALLY_ACHIEVED' then 0.5 end) partially_achieved, " +
+                    "sum(case when eosro.achievement_level = 'ACHIEVED' then 1 end) achieved_count " +
+                        "from referral ref " +
+                            "inner join end_of_service_report eosr on eosr.referral_id = ref.id " +
+                            "inner join end_of_service_report_outcome eosro on eosro.end_of_service_report_id = eosr.id " +
+                        "group by ref.id) sub), " +
+        "supplier_assessment_apointments AS ( " +
+            "select app.created_at as appointment_booked_time, " +
+            "case when (app.attended in ('LATE', 'YES')) then app.appointment_time else null end as attended_time, " +
+            "sup.referral_id " +
+                "from appointment app " +
+                    "join supplier_assessment_appointment saa on app.id = saa.appointment_id " +
+                    "join supplier_assessment sup on saa.supplier_assessment_id = sup.id ) " +
+        "select " +
+            "null AS referralLink, " +
+            "ref.reference_number AS referralReference, " +
+            "cast(ref.id as varchar) AS referralId, " +
```

Comment:
why do we have to do these casts?

Reply:
When using a native query, UUID's could not be automatically mapped and were throwing errors, therefore had to be cast